File: DailyChallenges/CPP/LeetCode_67.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        int na = a.size(), nb = b.size();
        if (na > nb) return addBinary(b, a);
        a = string(nb-na, '0') + a;
        int carrier = 0;
        for (int i = nb-1; i >=0 ; --i){
            carrier = (a[i]-'0') + (b[i]-'0') + carrier;
            b[i] = '0' + (carrier % 2);
            carrier = carrier / 2;
        }
        return ((carrier == 1)? "1":"") + b;
    }
};

File: DailyChallenges/Scala/LeetCode_67.scala
object Solution {
    def addBinary(a: String, b: String): String = {
        if (a.length > b.length) return addBinary(b, a);
        var l: List[Int] = Nil
        var c = 0
        for ((x, y) <- a.reverse.padTo(b.length, '0') zip b.reverse){
            c = c + x.asDigit + y.asDigit
            l =  (c%2) :: l
            c = c/2
        }
        if (c==1) l = 1 :: l
        l.mkString("")
    }
}
